src/ImageSharp/Processing/Processors/Binarization/AdaptiveThresholdProcessor.cs

```diff
@@ -46,80 +51,98 @@ public AdaptiveThresholdProcessor(TPixel upper, TPixel lower)
+                            {
+                                row[j].ToRgb24(ref rgb);
+                                sum += (ulong)(rgb.B + rgb.G + rgb.B);
+
+                                if (i != 0)
+                                {
+                                    intImage[i, j] = intImage[i - 1, j] + sum;
+                                }
+                                else
+                                {
+                                    intImage[i, j] = sum;
+                                }
+                            }
                         }
                     }
-                }
+                );
 
-                // How can I parallelize this?
-                ushort x1, x2, y1, y2;
+                ParallelHelper.IterateRows(
+                    workingnRectangle,
+                    configuration,
+                    rows =>
+                    {
+                        ushort x1, x2, y1, y2;
 
-                for (ushort i = startY; i < endY; i++)
-                {
-                    Span<TPixel> span = source.GetPixelRowSpan(i);
+                        uint count;
 
-                    for (ushort j = startX; j < endX; j++)
-                    {
-                        x1 = (ushort)Math.Max(i - s + 1, 0);
-                        x2 = (ushort)Math.Min(i + s + 1, endY - 1);
-                        y1 = (ushort)Math.Max(j - s + 1, 0);
-                        y2 = (ushort)Math.Min(j + s + 1, endX - 1);
+                        long sum;
 
-                        count = (uint)((x2 - x1) * (y2 - y1));
+                        for (int i = rows.Min; i < rows.Max; i++)
+                        {
+                            var row = source.GetPixelRowSpan(i);
 
-                        sum = intImage[x2, y2] - intImage[x1, y2] - intImage[x2, y1] + intImage[x1, y1];
+                            Rgb24 rgb = default;
 
-                        span[j].ToRgb24(ref rgb);
+                            for (int j = startX; j < endX; j++)
+                            {
+                                x1 = (ushort)Math.Max(i - s + 1, 0);
+                                x2 = (ushort)Math.Min(i + s + 1, endY - 1);
+                                y1 = (ushort)Math.Max(j - s + 1, 0);
+                                y2 = (ushort)Math.Min(j + s + 1, endX - 1);
 
-                        if ((rgb.R + rgb.G + rgb.B) * count < sum * (1.0 - 0.15))
-                        {
-                            span[j] = this.Lower;
-                        }
-                        else
-                        {
-                            span[j] = this.Upper;
+                                count = (uint)((x2 - x1) * (y2 - y1));
+
+                                sum = (long)(intImage[x2, y2] - intImage[x1, y2] - intImage[x2, y1] + intImage[x1, y1]);
+
+                                row[j].ToRgb24(ref rgb);
+
+                                if ((rgb.R + rgb.G + rgb.B) * count < sum * (1.0 - 0.15))
```

Comment:
this value represents **t** in the published paper. If the value of the current pixel is t percent less than this average of the s number pixels on the left and top then it is set to black, otherwise it is set to white.